build2 0.16.0 (new formula)

boris-kolpackov opened the pull request:
`build2` (https://build2.org) is a build system and package manager for C and C++, roughly like Cargo is for Rust.

Because it's a build system, `build2` uses multi-stage bootstrapping to build itself. This formula performs steps that are equivalent to invoking the upstream's `build.sh` script with the following options:

```
--local --private --system libpkgconf,libsqlite3
```

The only difference are a few customizations to find Homebrew's `libpkgconf` and `libsqlite3`. For background on these steps, see:

https://build2.org/build2-toolchain/doc/build2-toolchain-install.xhtml

The included test makes sure that a simple C++ project can be created, configured, and built using the newly-installed tools.

<!-- Use [x] to mark item done, or just click the checkboxes with device pointer -->

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

This is my attempt to resurrect PR #101846 by @tcbrindle . In particular, that PR was rejected because there was no way to build using Homebrew-provided `libpkgconf` (from `pkgconf` package) and `libsqlite3` (from `sqlite` package). This has been addressed in `build2` version `0.16.0`. I've also extended the formula's test a bit to exercise all the tools rather than only the build system as well as made other minor quality improvements (added `uses_from_macos` specifications, etc).

I've tested this on both Intel and M1 Mac OS. The build and test logs (attached) look nice and tidy to me.

[`build2-build.txt`](https://github.com/Homebrew/homebrew-core/files/12534460/build2-build.txt)

[`build2-test.txt`](https://github.com/Homebrew/homebrew-core/files/12534462/build2-test.txt)

boris-kolpackov commented:
I see the audit on macOS 11-arm64 failed like so:

```
brew audit build2 --online --new-formula
==> FAILED
Full audit build2 --online --new-formula output
  build2
    * Dependency 'sqlite' is provided by macOS; please replace 'depends_on' with 'uses_from_macos'.
```

This doesn't seem to be an issue on macOS 11-x86_64 nor on 12 and 13. At the same time looking at https://formulae.brew.sh/formula/sqlite it appears `sqlite` is available on Big Sur for both CPUs. Any ideas what might be going on here?

SMillerDev commented:
Well, it's exactly as the audit says. A version of SQLite is provided by macOS. Please replace the `depends_on "sqlite"` you have with a `uses_from_macos "sqlite"`

boris-kolpackov commented:
> Well, it's exactly as the audit says. A version of SQLite is provided by macOS.

Thanks but I must be not seeing the complete picture here. I actually want to depend on Homebrew's `sqlite` (because in the future it may be built with more suitable configuration; see https://github.com/Homebrew/homebrew-core/pull/101846#issuecomment-1131284823 for background). And I see quite a few existing formulas doing exactly that. For example, in `subversion@1.8.rb`:

```
depends_on "sqlite" # build against Homebrew version for consistency
```

I can also see that my local builds work as expects: `sqlite` package gets installed and `build2` gets built using Homebrew's `libsqlite3`. What am I missing here?

SMillerDev commented:
> I can also see that my local builds work as expects: sqlite package gets installed and build2 gets built using Homebrew's libsqlite3. What am I missing here?

It works fine, we just prefer not to depend on Homebrew software if the system provides a version.

> because in the future it may be built with more suitable configuration

In that case we can change the dependency in the future.

> And I see quite a few existing formulas doing exactly that.

If there is a good reason we can allow the audit failure, but I am not seeing a good reason here.

boris-kolpackov commented:
Ok, thanks for the clarification!

> If there is a good reason we can allow the audit failure, but I am not seeing a good reason here.

The overall reason is that `build2` has certain requirements on SQLite and while it's easy to know whether they are met with the Homebrew's version, it's unclear how Apple builds their versions nor are there any guarantees that they won't change things up in the future. Specifically:

1. `build2` need version 3.21.0 or later since it uses `ATTACH` support within transactions.

2. `build2` is sensitive to the `SQLITE_MAX_ATTACHED` built-time configuration value. We know that Homebrew's version is built with the default value `10` (which we hope can be increased in the future and the possibility of which was suggested in https://github.com/Homebrew/homebrew-core/pull/101846#issuecomment-1131833457). I have no idea what value Apple builds their versions with or that they won't decide to change them in the future.

Overall, I feel that using Homebrew's version will result in a less brittle user experience. But if you feel that this is still not good enough reason, I will go ahead and switch to the system version.

SMillerDev commented:
> build2 need version 3.21.0 or later since it uses ATTACH support within transactions.

The oldest supported macOS version has SQLite 3.32.3 so that should be fine.

> build2 is sensitive to the SQLITE_MAX_ATTACHED built-time configuration value. We know that Homebrew's version is built with the default value 10. I have no idea what value Apple builds their versions with or that they won't decide to change them in the future.

Such guarantees do not exist for the Homebrew version either, we can always switch if a lot of users have issues with this.

SMillerDev commented on Formula/b/build2.rb:
```diff
@@ -0,0 +1,240 @@
+           "-j", ENV.make_jobs,
+           "install:", *modules.map { |m| "#{m}/" }
+
+    # Verify all the modules can be loaded by loading their tests.
+    #
+    system "b", "-v",
+           "noop:", *modules.map { |m| "tests/#{m}-tests/" }
```
That should be in the test block

boris-kolpackov replied:
I don't believe this will be straightforward since we don't have access to source code during tests (and which is used to get the list of modules). But seeing that we already try to load one of the modules (`autoconf`) in tests, I can just drop this.

boris-kolpackov replied:
Resolved in [44e1301](https://github.com/Homebrew/homebrew-core/pull/141550/commits/44e130191b977fa3b007d8dae23578aa42358695).

SMillerDev commented on Formula/b/build2.rb:
```diff
@@ -0,0 +1,240 @@
+
+    ENV.prepend_path "PATH", bin
+
+    system "command", "-v", "b"
+    system "command", "-v", "bpkg"
+    system "command", "-v", "bdep"
+
+    system "b",    "--version"
+    system "bpkg", "--version"
+    system "bdep", "--version"
```
This shouldn't be here

boris-kolpackov replied:
These are just some sanity checks (to make sure we don't pick up an unrelated toolchain from a different location). But I will drop them.

boris-kolpackov replied:
Resolved in [44e1301](https://github.com/Homebrew/homebrew-core/pull/141550/commits/44e130191b977fa3b007d8dae23578aa42358695).

boris-kolpackov commented:
Ok, I've switched to the Mac OS SQLite version. I've retested the formula on Intel and M1 Mac OS and also confirmed `/usr/lib/libsqlite3.dylib` is now used.

Now I get the following CI audit error:

```
Full audit build2 --online --new-formula output
  build2
    * Formulae should not require patches to build. Patches should be submitted and accepted upstream first.
```

The patches have been submitted and accepted upstream, specifically: https://github.com/build2/build2/commit/f221ca94a304abaa79239cf2cbaf3ffafbbc092c and https://github.com/build2/build2/commit/17c1214943959b304ab5cdcccacec8a70c7a6c0f . Not sure what I should do next.

boris-kolpackov commented on Formula/b/build2.rb:
```diff
@@ -0,0 +1,240 @@
+  end
+end
+
+# We have to adjust the file paths in the below patches to match the source
+# archive we are using (e.g., a/libbuild2/... -> a/build2/libbuild2/...).
+#
```
Hm, thanks for the tip but I couldn't find any documentation/examples of this and trying something like:

```ruby
  patch do
    path "build2/"
    url "https://github.com/build2/build2/commit/f221ca94a304abaa79239cf2cbaf3ffafbbc092c.patch?full_index=1"
    sha256 "391fbd38e863bc3c23f1cddf360f097d210b31ad1e9a906eb583d927f0d216a3"
  end
```

Gives:

```
Error: build2: undefined method `path' for #<Resource::PatchResource:0x00007fa6ef520600>
Did you mean?  paths
               patch
```

boris-kolpackov replied:
Ok, looks like the correct syntax is:

```ruby
  patch do
    url "https://github.com/build2/build2/commit/f221ca94a304abaa79239cf2cbaf3ffafbbc092c.patch?full_index=1"
    sha256 "391fbd38e863bc3c23f1cddf360f097d210b31ad1e9a906eb583d927f0d216a3"
    directory "build2"
  end
```

boris-kolpackov replied:
Fixed in [94b05da](https://github.com/Homebrew/homebrew-core/pull/141550/commits/94b05dafd38cd5a144b98fe711065d8b829d5757).

SMillerDev commented:
For the patches, let's just wait for the next release.

boris-kolpackov commented:
> For the patches, let's just wait for the next release.

That's disappointing, especially seeing that we are waiting from the `0.14.0` release.

I appreciate the desire to avoid introducing new formulas that require patches, but perhaps an exception can be made seeing that we are dealing with a C/C++ build system, which is notoriously prone to corner cases? Specifically, both patches try to deal with defects and breaking changes in Mac OS. The first one compensates for the lack of SDK path in the `-print-search-dirs` report printed by Apple clang. The second deals with new invisible/inaccessible `.dylib` libraries in `/usr/lib/`. Chances are we will have to work around something else come next release.

Alternatively, we could drop the patches if we switch back to Homebrew SQLite. I could switch back to Mac OS SQLite after the release.

What do you think?

boris-kolpackov commented:
And another point perhaps worth mentioning: we are getting ready to release the new version of `xsd` which now uses `build2` as the build system. Currently the [`xsd.rb`](https://github.com/Homebrew/homebrew-core/blob/master/Formula/x/xsd.rb) formula includes a bunch of patches. After the release they will be all gone. So by temporarily adding a few patches here we are going to get rid of some elsewhere.

boris-kolpackov commented:
@SMillerDev could I please have some guidance here before I go on a wild goose chase? Specifically, I am hearing contradictory requirements and I am not sure which one has priority. On the one hand, it appears that keeping the formula as simple and easy to understand as possible is key for acceptance ("this one is already complicated enough without the patches that I don't feel comfortable shipping it with them"). On the other, the desire to use Mac OS SQLite on Mac OS but having to still use Homebrew SQLite on Linux will complicate the formula even further because the use of Homebrew SQLite requires getting the header/library search paths from the formula and passing them to the build system. So we will have two different ways of building things for the two platforms.

On the other hand, if we go with Homebrew SQLite everywhere (as I originally tried), then the formula will be nice and simple and we can also drop the patches.

SMillerDev commented:
> On the one hand, it appears that keeping the formula as simple and easy to understand as possible is key for acceptance. 

It is
> On the other, the desire to use Mac OS SQLite on Mac OS but having to still use Homebrew SQLite on Linux will complicate the formula even further because the use of Homebrew SQLite requires getting the header/library search paths from the formula and passing them to the build system. So we will have two different ways of building things for the two platforms.

There are multiple ways for build systems to detect where dependencies are installed, and even if you have to hardcode the paths I would hardly classify them as "different ways of building things". 

> On the other hand, if we go with Homebrew SQLite everywhere (as I originally tried), then the formula will be nice and simple and we can also drop the patches.

The state of the art on macOS means that we can use system components, so we should since that means the dependency tree is smaller. On Linux we have no guarantees about what will be installed so we have to use the versions we ship.

boris-kolpackov commented:
Ok, thanks, that makes the direction clear. Will try to fix the Linux build shortly.

> There are multiple ways for build systems to detect where dependencies are installed [...]

Hm, for SQLite specifically, neither the library nor the `pkg-config` file are installed in any standard locations (e.g., `/usr/local/lib/`) where the build system can be expected to look by default. So short of hardcoding Homebrew-specific installation locations (which are, BTW, different for Intel and M1), I don't see how a build system can automatically detect where `libsqlite3` is installed. Or am I missing something here?

SMillerDev commented:
> Hm, for SQLite specifically, neither the library nor the pkg-config file are installed in any standard locations (e.g., /usr/local/lib/) where the build system can be expected to look by default. 

But the `pkg-config` that Homebrew ships does know where to look, try: `pkg-config --variable pc_path pkg-config`

boris-kolpackov commented:
> But the `pkg-config` that Homebrew ships does know where to look [...]

Hm, it doesn't seem to work for `sqlite`:

```
$ brew install sqlite pkg-config

$ pkg-config --variable pc_path pkg-config
/usr/local/lib/pkgconfig:/usr/local/share/pkgconfig:/usr/lib/pkgconfig:/usr/local/Homebrew/Library/Homebrew/os/mac/pkgconfig/13

$ pkg-config --cflags --libs sqlite3
-lsqlite3
```

In particular, of the printed paths the only one that contains `sqlite3.pc` is the last (`/usr/local/Homebrew/Library/Homebrew/os/mac/pkgconfig/13`).

SMillerDev commented:
> Hm, it doesn't seem to work for sqlite

Doesn't the `$ pkg-config --cflags --libs sqlite3` run show that it does actually work?

boris-kolpackov commented:
> Doesn't the `$ pkg-config --cflags --libs sqlite3` run show that it does actually work?

I guess it depends on how you define "works". If you define it as "always use Mac OS `sqlite` regardless of the presence of the Homebrew `sqlite` package", then yes.

But your original suggestion was in response to my statement that a build system cannot possibly find Homebrew `sqlite` automatically short of hardcoding Homebrew-specific installation paths. And in response to that statement the currently behavior definitely doesn't work.

SMillerDev commented:
> I guess it depends on how you define "works". If you define it as "always use Mac OS sqlite regardless of the presence of the Homebrew sqlite package", then yes.

I define it as "can find the SQLite install". That it'll always give the one you should use in Homebrew is just a nice side-effect.
> But your original suggestion was in response to my statement that a build system cannot possibly find Homebrew sqlite automatically short of hardcoding Homebrew-specific installation paths. And as the response to that statement the current behavior definitely doesn't work.

It finds the one it needs right? On Linux I think it will find the brew one.

boris-kolpackov commented:
> I define it as "can find the SQLite install". That it'll always give the one you should use in Homebrew is just a nice side-effect.

If one should always use Mac OS `sqlite` then why have a Homebrew version in the first place? There are probably good reasons for having a Homebrew version since currently there are 56 packages that have `depends_on "sqlite"`. For example, `deno.rb` has this comment: `# needs `sqlite3_unlock_notify` so there are clearly situations where the Apple-built `sqlite` is not appropriate. And if that's the case, then for `pkg-config` always returning the Mac OS version is a not-always-desirable behavior.

> It finds the one it needs right? On Linux I think it will find the brew one.

I guess it depends on who "it" refers to. If it's a package like `deno.rb` that wishes to always use Homebrew version, then no, it doesn't find what it needs.

In any case, I also appreciate that this won't be easy to fix short of patching `pkg-config` to signal which version is desired. So the current behavior is probably the best it can be given the decades of baggage we have accumulated in this area.

SMillerDev commented on Formula/b/build2.rb:
```diff
@@ -0,0 +1,143 @@
+           "config.cc.coptions=-O3",
+           "config.cc.loptions=#{loptions}",
+           "config.bin.lib=shared",
+           "config.bin.rpath=#{lib}/build2",
```
```suggestion
           "config.bin.rpath=#{rpath}",
```

boris-kolpackov replied:
I don't believe this will work unless we also get rid of the private installation (`config.install.private=build2`). Here is what happens: `build2` consists of a bunch of `.dylib` libraries which are build system modules. These libraries are not meant to be linked by other applications and not to pollute `$prefix/lib/` we install them into `$prefix/lib/build2/` and arrange for the corresponding `rpath`. This is also how our standard installation script installs things.

I can change that (i.e., install all the libraries directly into `$prefix/lib/`) but I don't see any benefits in doing that. The current layout works without any issues.

boris-kolpackov replied:
Ok, I see what `#{rpath}` tries to accomplish (`@loader_path/../lib`). Let me see how I can work it in.

boris-kolpackov replied:
Fixed in https://github.com/Homebrew/homebrew-core/pull/141550/commits/a97e5b511a89af6531e68b69f15e9ae6ea61d2bf , thanks for pointing this out!

SMillerDev commented on Formula/b/build2.rb:
```diff
@@ -0,0 +1,143 @@
+
+    system "build2/build2/b-boot", "-V",
+           "config.cxx=#{ENV.cxx}",
+           "config.cc.coptions=-O3",
```
I don't think Homebrew respects this, it'll be stripped out.

boris-kolpackov replied:
You mean `-O3`? I checked the build logs and the option appears to be passed to `clang`. But I can change it to something else if there is something more appropriate.

boris-kolpackov replied:
Ah, I see, thanks. I would need to still keep them since they will be included into configurations that will be used by the user to build build-time dependencies. I've added the following note to the formula:

```
    # Note that while Homebrew's clang wrapper will strip any optimization
    # options, we still want to pass them since they will also be included
    # into the ~host and ~build2 configurations that will be used to build
    # built-time dependencies and build system modules, respectively, when
    # the user uses actual clang.
```

boris-kolpackov commented:
The change to `#{rpath}` cause the Linux build to fail but Mac OS is ok. The relevant diagnostics from the Linux CI:

```
2023-09-12T11:01:20.8105805Z [34m==>[0m [1mbrew linkage --test build2[0m
2023-09-12T11:01:22.5034194Z [31m==>[0m [1m[31mFAILED[0m[0m
2023-09-12T11:01:22.5038980Z ##[group]Full linkage --test build2 output
2023-09-12T11:01:22.5039308Z Missing libraries:
2023-09-12T11:01:22.5039690Z   unexpected (libbuild2-0.16.so)
2023-09-12T11:01:22.5040059Z   unexpected (libbuild2-bin-0.16-0.16.so)
2023-09-12T11:01:22.5040431Z   unexpected (libbuild2-cc-0.16-0.16.so)
2023-09-12T11:01:22.5041171Z   unexpected (libbuild2-cxx-0.16-0.16.so)
2023-09-12T11:01:22.5041548Z   unexpected (libbuild2-in-0.16-0.16.so)
2023-09-12T11:01:22.5041884Z   unexpected (libbutl-0.16.so)
2023-09-12T11:01:22.5042202Z   unexpected (libodb-2.5.0-b.25.so)
2023-09-12T11:01:22.5042857Z ##[endgroup]
```

And later I also see this:

```
2023-09-12T11:01:33.9199720Z ==> brew linkage --cached build2
2023-09-12T11:01:33.9200378Z System libraries:
2023-09-12T11:01:33.9200917Z   /lib/x86_64-linux-gnu/libc.so.6
2023-09-12T11:01:33.9202243Z   /lib/x86_64-linux-gnu/libgcc_s.so.1
2023-09-12T11:01:33.9202742Z   /lib/x86_64-linux-gnu/libstdc++.so.6
2023-09-12T11:01:33.9204014Z Homebrew libraries:
2023-09-12T11:01:33.9204497Z   /home/linuxbrew/.linuxbrew/Cellar/build2/0.16.0/bin/../lib/build2/libbpkg-0.16.so (build2)
2023-09-12T11:01:33.9208697Z   /home/linuxbrew/.linuxbrew/Cellar/build2/0.16.0/bin/../lib/build2/libbuild2-0.16.so (build2)
2023-09-12T11:01:33.9209273Z   /home/linuxbrew/.linuxbrew/Cellar/build2/0.16.0/bin/../lib/build2/libbuild2-bash-0.16-0.16.so (build2)
2023-09-12T11:01:33.9229637Z   /home/linuxbrew/.linuxbrew/Cellar/build2/0.16.0/bin/../lib/build2/libbuild2-bin-0.16-0.16.so (build2)
2023-09-12T11:01:33.9230282Z   /home/linuxbrew/.linuxbrew/Cellar/build2/0.16.0/bin/../lib/build2/libbuild2-c-0.16-0.16.so (build2)
2023-09-12T11:01:33.9230849Z   /home/linuxbrew/.linuxbrew/Cellar/build2/0.16.0/bin/../lib/build2/libbuild2-cc-0.16-0.16.so (build2)
2023-09-12T11:01:33.9231416Z   /home/linuxbrew/.linuxbrew/Cellar/build2/0.16.0/bin/../lib/build2/libbuild2-cli-0.16-0.16.so (build2)
2023-09-12T11:01:33.9231985Z   /home/linuxbrew/.linuxbrew/Cellar/build2/0.16.0/bin/../lib/build2/libbuild2-cxx-0.16-0.16.so (build2)
2023-09-12T11:01:33.9232544Z   /home/linuxbrew/.linuxbrew/Cellar/build2/0.16.0/bin/../lib/build2/libbuild2-in-0.16-0.16.so (build2)
2023-09-12T11:01:33.9233115Z   /home/linuxbrew/.linuxbrew/Cellar/build2/0.16.0/bin/../lib/build2/libbuild2-version-0.16-0.16.so (build2)
2023-09-12T11:01:33.9233653Z   /home/linuxbrew/.linuxbrew/Cellar/build2/0.16.0/bin/../lib/build2/libbutl-0.16.so (build2)
2023-09-12T11:01:33.9234184Z   /home/linuxbrew/.linuxbrew/Cellar/build2/0.16.0/bin/../lib/build2/libodb-2.5.0-b.25.so (build2)
2023-09-12T11:01:33.9234755Z   /home/linuxbrew/.linuxbrew/Cellar/build2/0.16.0/bin/../lib/build2/libodb-sqlite-2.5.0-b.25.so (build2)
2023-09-12T11:01:33.9235182Z   /home/linuxbrew/.linuxbrew/opt/pkgconf/lib/libpkgconf.so.4 (pkgconf)
2023-09-12T11:01:33.9235569Z   /home/linuxbrew/.linuxbrew/opt/sqlite/lib/libsqlite3.so.0 (sqlite)
2023-09-12T11:01:33.9235868Z Missing libraries:
2023-09-12T11:01:33.9236125Z   libbuild2-0.16.so
2023-09-12T11:01:33.9236413Z   libbuild2-bin-0.16-0.16.so
2023-09-12T11:01:33.9236712Z   libbuild2-cc-0.16-0.16.so
2023-09-12T11:01:33.9237008Z   libbuild2-cxx-0.16-0.16.so
2023-09-12T11:01:33.9237290Z   libbuild2-in-0.16-0.16.so
2023-09-12T11:01:33.9237561Z   libbutl-0.16.so
2023-09-12T11:01:33.9237819Z   libodb-2.5.0-b.25.so
```

Not sure what's going here. I believe those `.../bin/../lib/...` paths indicate that this is a relocatable installation (`#{rpath}` is probably `$ORIGIN/../lib`). The libraries seem to be there but for some reason `brew` seems to think they are missing. The same works fine if not using `#{rpath}`.

@SMillerDev any ideas what might be going on here?

boris-kolpackov commented:
Ok, I've retested everything regarding this Linux linkage issue and here is the summary of the situation:

* private install without `#{rpath}` -- works

* normal install with `#{rpath}` -- works

* private install with `#{rpath}` -- fails

Here "private install" means libraries get installing in `#{lib}/build2` and "normal install" means libraries get installed into `#{lib}`. And "with `#{rpath}`" means the `-rpath` option is based on `#{rpath}` (which I assume is `$ORIGIN/../lib`) and "without `#{rpath}`" means the `-rpath` option is based on `#{lib}`.

Looks like a Homebrew bug to me. @SMillerDev thoughts? Should I report this?

boris-kolpackov commented:
I've filed an issue with all the details: https://github.com/Homebrew/brew/issues/16001

boris-kolpackov commented:
Ok, my misunderstanding has been corrected in the above issue and I have the final (from my POV, at least) version of the formula with a green CI. Key differences compared to the previous review:

1. I have disabled private installation to keep things simple and the formula now uses `#{rpath}`.

2. I found a simple work around that allowed me to build everything without any patching. So the formula is now patch-free. (And it still uses system `sqlite` on Mac OS and Homebrew `sqlite` on Linux, as requested.)

SMillerDev commented on Formula/b/build2.rb:
```diff
@@ -0,0 +1,149 @@
+  sha256 "23793f682a17b1d95c80bbd849244735ed59a3e27361529aa4865d2776ff8adc"
+  license "MIT"
+
+  # TODO: add livecheck based on https://download.build2.org/toolchain.sha256
+  #       for background, see https://download.build2.org/README
+
```
Still TODO

boris-kolpackov replied:
Added.

SMillerDev commented on Formula/b/build2.rb:
```diff
@@ -0,0 +1,149 @@
+  # TODO: add livecheck based on https://download.build2.org/toolchain.sha256
+  #       for background, see https://download.build2.org/README
+
+  # Note that for now we use SQLite from Mac OS instead of Homebrew at the
+  # insistence of Homebrew maintainers but may switch in the future if have a
+  # good enough reason (for example, higher SQLITE_MAX_ATTACHED limit). See PR
+  # #141550 for details. Note also that on Linux it comes from Homebrew so
+  # we have to arrange for library search paths, etc (see below).
+  #
```
This is all in the PR, no need to have it in the file as well

boris-kolpackov replied:
Dropped.

SMillerDev commented on Formula/b/build2.rb:
```diff
@@ -0,0 +1,149 @@
+    pkgconf = Formula["pkgconf"]
+    sqlite = Formula["sqlite"] if OS.linux?
+
+    # Note that the below steps are equivalent to invoking build.sh with the
+    # --local --system libpkgconf,libsqlite3 options. The only difference are
+    # a few customizations to find Homebrew's libpkgconf and libsqlite3.
+    #
+    # For background on these steps, see:
+    #
+    # https://build2.org/build2-toolchain/doc/build2-toolchain-install.xhtml
+
+    # Suppress loading of default options files.
+    #
```
```suggestion
    # Suppress loading of default options files.
```

boris-kolpackov replied:
Applied, thanks.

SMillerDev commented on Formula/b/build2.rb:
```diff
@@ -0,0 +1,149 @@
+    # Note that we disable all warnings since we cannot do anything more
+    # granular during bootstrap stage 1.
+    #
+    cd "build2" do
```
why not stay in the `build2` directory with `chdir`?

boris-kolpackov replied:
Replaced all cases of `cd` with `chdir`, thanks for the tip!

SMillerDev commented on Formula/b/build2.rb:
```diff
@@ -0,0 +1,149 @@
+      system "make",
+             "-j", ENV.make_jobs,
+             "-f", "./bootstrap.gmake",
+             "CXX=#{ENV.cxx}", "CXXFLAGS=-w"
```
Does it not work when it just tries to detect this from the system?

boris-kolpackov replied:
Yes, indeed, now simplified, thanks!

SMillerDev commented on Formula/b/build2.rb:
```diff
@@ -0,0 +1,149 @@
+             "CXX=#{ENV.cxx}", "CXXFLAGS=-w"
+    end
+
+    # Bootstrap, stage 2.
+    #
+    # Force using Homebrew's libpkgconf rather than building libpkg-config
+    # from source.
+    #
+    # Note that the build2 driver bootstrapped at stage 1 above doesn't read
+    # .pc files. Thus, during the bootstrap stage 2, it is assumed that the
+    # headers and libraries of the system dependencies can be found by the
+    # compiler and linker automatically. Failed that we have to manually
+    # specify their header/library paths, which is what we do here for
+    # libpkgconf.
+    #
```
```suggestion
```

boris-kolpackov replied:
Applied, thanks!

SMillerDev commented on Formula/b/build2.rb:
```diff
@@ -0,0 +1,149 @@
+      mv "build2/b", "build2/b-boot"
+    end
+
+    # Local build.
+    #
+    # Note that we don't build hermetically to avoid baking Homebrew-specific
+    # paths into the final dylibs.
+    #
+    # Note also that we currently disable private installation since combined
+    # with #{rpath} it causes `brew linkage` failures. See brew issue #16001
+    # for the reasons. The easiest solution is probably to enable relocatable
+    # installation (config.install.relocatable=true) on our side which will
+    # additionally produce relocatable pkg-config files. But for now let's
+    # not complicate things.
+    #
+    loptions = "-L#{pkgconf.lib}"
+    loptions += " -L#{sqlite.lib}" if OS.linux?
+
+    # Work around invisible/inaccessible /usr/lib/libsqlite3.dylib.
+    # Specifically, on recent versions of Mac OS, /usr/lib/ is no longer
+    # accessible and in order to link to system libraries one has to go
+    # through .tbd files ("text-based stub libraries") found in
+    # .../MacOS*.sdk/usr/lib/.
+    #
+    # Note: drop this workaround once upgraded to build2 0.17.0 which can
+    # handle .tbd files directly (see commits 391fbd38 and a860ce05).
+    #
+    if OS.mac?
+      mkdir "#{buildpath}/build/lib"
+      ln_s "#{MacOS.sdk_path}/usr/lib/libsqlite3.tbd", "#{buildpath}/build/lib/libsqlite3.dylib"
+      loptions += " -L#{buildpath}/build/lib"
+    end
```
```suggestion
    loptions = "-L#{pkgconf.lib}"
    loptions += " -L#{sqlite.lib}" if OS.linux?

    # Work around macOS dylib cache in /usr/lib
    if OS.mac?
      mkdir "#{buildpath}/build/lib"
      ln_s "#{MacOS.sdk_path}/usr/lib/libsqlite3.tbd", "#{buildpath}/build/lib/libsqlite3.dylib"
      loptions += " -L#{buildpath}/build/lib"
    end
```

boris-kolpackov replied:
Applied, thanks!

SMillerDev commented on Formula/b/build2.rb:
```diff
@@ -0,0 +1,149 @@
+           "config.cc.loptions=#{loptions}",
+           "config.bin.lib=shared",
+           "config.bin.rpath='#{rpath}'",
+           # "config.bin.rpath='#{lib}/build2'",
```
```suggestion
```

boris-kolpackov replied:
Applied, thanks!

SMillerDev commented on Formula/b/build2.rb:
```diff
@@ -0,0 +1,149 @@
+           "config.bin.rpath='#{rpath}'",
+           # "config.bin.rpath='#{lib}/build2'",
+           "config.install.root=#{prefix}",
+           # "config.install.private=build2",
```
```suggestion
```

boris-kolpackov replied:
Applied, thanks!

SMillerDev commented on Formula/b/build2.rb:
```diff
@@ -0,0 +1,149 @@
+
+    ENV.prepend_path "PATH", bin
+
+    # Install build system modules.
+    #
+    modules = buildpath.glob("libbuild2-*/").map(&:basename)
+
+    system "b", "-v",
+           "-j", ENV.make_jobs,
+           "install:", *modules.map { |m| "#{m}/" }
```
I think this should work like this too
```suggestion
    system "b", "-v",
           "-j", ENV.make_jobs,
           "install:", *Dir["libbuild2-*/"]
```

boris-kolpackov replied:
Yes, applied, thanks!

SMillerDev commented on Formula/b/build2.rb:
```diff
@@ -0,0 +1,149 @@
+    cd "libhello" do
+      system bin/"bdep", "init", "--config-create", "@default", "cc"
+      system bin/"b", "test"
+      system bin/"b", "-V", "noop:", "libhello/" # Show configuration report.
```
Can we assert some output here?

boris-kolpackov replied:
Added for tool versions as well as project info.

SMillerDev commented on Formula/b/build2.rb:
```diff
@@ -0,0 +1,106 @@
+
+  test do
+    # Suppress loading of default options files.
+    #
```
```suggestion
```

boris-kolpackov replied:
Applied, thanks!

SMillerDev commented on Formula/b/build2.rb:
```diff
@@ -0,0 +1,106 @@
+
+    # Note that we disable all warnings since we cannot do anything more
+    # granular during bootstrap stage 1.
+    #
```
```suggestion
```

boris-kolpackov replied:
Applied, thanks!

SMillerDev commented on Formula/b/build2.rb:
```diff
@@ -0,0 +1,106 @@
+  sha256 "23793f682a17b1d95c80bbd849244735ed59a3e27361529aa4865d2776ff8adc"
+  license "MIT"
+
+  # For background, see https://download.build2.org/README
+  #
```
Let's just remove this comment, it doesn't make anything clearer for me.

boris-kolpackov replied:
Removed.

boris-kolpackov commented:
@SMillerDev thanks for the prompt review! I've addressed all the issues and I expect the CI to finish green seeing that the last batch are comment-only changes (the ~1h CI turnaround is truly torturous).

boris-kolpackov commented:
@SMillerDev Please let me know if everything is ok now or if I need to make any other changes.

chenrui333 commented:
LGTM

( @boris-kolpackov thanks for your first PR to homebrew-core!)

SMillerDev reviewed (approved):
Thanks @boris-kolpackov  ! Without contributions like yours it'd be impossible to keep homebrew going with the high standards that users have come to expect from the project. You can feel good knowing that you've made the world a tiny bit better for homebrew users around the world! 👍 🎉

boris-kolpackov commented:
Thanks @SMillerDev and @chenrui333!

If there are any issues with this formula in the future, please ping me. I also intend to maintain it going forward.

boris-kolpackov commented:
@SMillerDev , @chenrui333 Is there anything else I need to do to have this merged? This PR has been sitting in the approved state for 5 days now I am not sure if this is normal (though I see new formulas being merges that we approved less than a day ago) or something still needs to be done.

p-linnane reviewed (approved):
Thanks for your patience and contribution @boris-kolpackov.

p-linnane commented:
@boris-kolpackov Can you squash all your commits into one with the message `build2 0.16.0 (new formula)`? This will make the merge process easier. Thanks!

boris-kolpackov commented:
@p-linnane
> Can you squash all your commits into one [...]

 Done, the CI is green. Thanks for looking into this and let me know if there is anything else.

p-linnane reviewed (approved):
For some reason GitHub's API says this PR is not approved, so trying to approve it again.

p-linnane commented:
@boris-kolpackov Thanks for your patience! This PR had so many comments that the GitHub API default pagination was cutting off before the approvals. This should be fixed in https://github.com/Homebrew/homebrew-core/pull/147313, so we should be able to merge this shortly.

boris-kolpackov commented:
Thank you everyone for your help on getting this through, much appreciated!
